### ISSUE TITLE
Make hvPlot tooltips useful in modern editors and IDEs

### ISSUE THREAD

**Issue by MarcSkovMadsen** [holoviz/hvplot#789]
## Request

Make sure tooltips shown in modern Editors and IDEs like VS Code and PyCharm are available and useful.

## Motivation

I'm trying to help a user of hvPlot here https://discourse.holoviz.org/t/linked-plot-is-exceedingly-slow/3966. I am a user of VS Code. Its just a very slow process helping the user because I need to search and read things on the hvPlot website instead of being able to stay in my editor.

The issue is that the tooltips shown by VS Code are either empty or not useful. My guess is that its because no one has ever considered to make sure the docstrings of hvPlot works in modern editors and IDEs.

The consequence is that hvPlot is much, much harder to learn and work with than it has to be for editor and IDE users.

## scatter

`.hvplot.scatter` has no tooltip.

![image](https://user-images.githubusercontent.com/42288570/180633157-7e39aacf-78e5-4829-bad6-f6e0ced3b672.png)

**Comment by maximlt** [comment]
I agree that better docstrings and documentation is required.

> I am a user of VS Code. Its just a very slow process helping the user because I need to search and read things on the hvPlot website instead of being able to stay in my editor.

Of course the more you can stay working in your editor the better. However don't forget that because of the flat API of the `.plot()` API of Pandas, which applies to the `.hvplot()` API of hvPlot, the provided methods accept a large number of arguments, and this by design. See for instance the tooltip of Pandas `df.plot.scatter()`, which doesn't list all the kwargs it accepts.

![image](https://user-images.githubusercontent.com/35924738/180636880-ed2b58ce-c0e9-4a6e-bdb4-17c6a84601b5.png)

In most cases then you will have to go to Pandas website and look for what kwargs you can pass. The same applies to hvPlot, even more so as it provides more powerful arguments for which a simple explanation in a tooltip may not be enough to be understood properly (e.g. `rasterize`).

Also note that Pandas docstring doesn't link to the reference page and the *Examples* section doesn't display plots. Even projects like Pandas with about 100M downloads/month have difficulties to make this right.

![image](https://user-images.githubusercontent.com/35924738/180637132-ad1d684c-326a-43f2-8d39-315562c0285a.png)

hvPlot and I think the HoloViz tools in general have been optimized to work well in Jupyter Notebooks. Jupyter Notebooks are also modern editors, and this is where I expect most users currently are. hvPlot docstrings work better there, even more so as they are dynamically adapted the activated backend (e.g. bokeh or matplotlib or plotly) to provide an appropriate list of styling options.

![image](https://user-images.githubusercontent.com/35924738/180637107-5183dcc2-0cfc-4209-bcff-d71478c2358f.png)

There are of course users like you who use hvPlot to write code in a Python script in VSCode or PyCharm, and I expect more and more people will use the notebook support offered by VSCode. The experience of these users should be taken more into account.

**Comment by MarcSkovMadsen** [comment]
Thanks @maximlt . I agree with the above. I will try to make some example PRs to show what can be done, how and the value it creates.

**Comment by maximlt** [comment]
Ok cool! Make sure it doesn't affect the experience of the Jupyter notebooks users though 🙃

**Comment by MarcSkovMadsen** [comment]
Regarding Matplotlib documentation. They have the advantage of huge user base and community. So they don't need to get their documentation nearly as right as HoloViz do. Matplotlib is a standard and benchmark that people will spend the effort to learn just because its recommend and documented in many, many places. hvPlot and HoloViews is starting in a different place.

**Comment by MarcSkovMadsen** [comment]
I will try to keep the Jupyter experience. The problem is that I cannot install jupyterlab and test as conda simply cannot resolve the environment for jupyterlab. At least on Python 3.6 that I installed as described in the developer docs.

![image](https://user-images.githubusercontent.com/42288570/180639197-be5e3d77-a61b-40e6-81b9-d34c4369f1d6.png)

**Comment by maximlt** [comment]
Well if you've pip installed hvPlot then you might as well just install jupyterlab with pip.

**Comment by MarcSkovMadsen** [comment]
Yep. Gave up and just made a normal virtual environment and used `pip`. It works.

**Comment by MarcSkovMadsen** [comment]
I've tried jupyter lab for hvplot with both TAB and SHIFT+TAB completion but I don't get any nice help as you show above. How do I get what you see @maximlt ?

![image](https://user-images.githubusercontent.com/42288570/180639878-427027b8-3303-4b17-9384-c711c6628132.png)

**Comment by maximlt** [comment]
Is `df` really a Pandas DataFrame in your code snippet?

And yes that was with `Shift + Tab`.

**Comment by MarcSkovMadsen** [comment]
My mistake. Thanks.

**Comment by MarcSkovMadsen** [comment]
I've added some example module level docstrings. But hvplot will not be easy to get nicely working in vs code because of the way its implemented using monkey patching. Static type checkers do not understand that. So right now I cannot see how we will ever be able to provide tooltips for `df.hvplot` or `df.hvplot.scatter`.

The alternative would be to a add a description on how to best work with hvplot in VS Code and other IDEs to be able to navigate and use hvplot efficiently.

**Comment by maximlt** [comment]
Yep right, I don't tools like Pylance can now cover that, and I'm also not sure that these tools cover the case when a library uses the API offered by Pandas or Xarray to register an accessor.

What we could say to those users is that instead of calling `df.hvplot()` they can directly use the `hvPlot` class to wrap the dataframe and then will get autocompletion working in a script:
![image](https://user-images.githubusercontent.com/35924738/180647280-c3adaff6-ba8c-45dc-9ff6-d680e8571cb2.png)

I have duplicated suggestions in the screenshot above, it might be my environment that is partially broken. Do you also get these duplicates?

**Comment by MarcSkovMadsen** [comment]
Yeah. I think `hvPlot` could be useful. I was in doubt whether it just worked on Pandas. But it seems to work on more data sources. Here its xarray.

![image](https://user-images.githubusercontent.com/42288570/180698724-31960c50-b5da-4f0f-a1c7-68df6c0b1ad0.png)
